Feat(docs): Added preset to Date Range Picker

Shreypatel13ll opened the pull request:
1. **Internal Presets:** Integrated internal presets into the date-range picker, accessible through buttons.

2. **External Presets:** Introduced external presets for the date-range picker, available via a select popover.

3. **Changed Arrangement:**
    - Date Picker
    - Date Picker with Internal Presets
    - Date-Range Picker
    - Date-Range Picker with Internal Presets
    - Date-Range Picker with External Presets
    
### Images
- ![image](https://github.com/shadcn-ui/ui/assets/68627196/9c08dc78-5816-4600-9da9-347805650454)

- ![image](https://github.com/shadcn-ui/ui/assets/68627196/1f861f62-bb0c-4635-9485-9c5bb0a499b3)

Koning3ton commented:
Great extension! But is it not a good idea to integrate the different options into one component?

### usage example:
```javascript
<CalendarDateRangePicker showExternalPresets showInternalPresets numberOfMonths={3} />
````
### Component:
```javascript
'use client';

import * as React from 'react';
import { CalendarIcon } from '@radix-ui/react-icons';
import { addDays, format } from 'date-fns';
import { DateRange } from 'react-day-picker';

import { cn } from '@/lib/utils';
import { Button } from '@/components/ui/button';
import { Calendar } from '@/components/ui/calendar';
import { Popover, PopoverContent, PopoverTrigger } from '@/components/ui/popover';
import { Select, SelectContent, SelectItem, SelectTrigger, SelectValue } from '@/components/ui/select';

interface CalendarDateRangePickerProps extends React.HTMLAttributes<HTMLDivElement> {
	showExternalPresets?: boolean;
	showInternalPresets?: boolean;
	numberOfMonths?: 2 | 3 | 4 | 5 | 6;
}

export function CalendarDateRangePicker({ showExternalPresets, showInternalPresets, numberOfMonths = 2, className }: CalendarDateRangePickerProps) {
	const [date, setDate] = React.useState<DateRange | undefined>({
		from: new Date(2023, 0, 20),
		to: addDays(new Date(2023, 0, 20), 20)
	});

	const handlePresetSelect = (value: string) => {
		switch (value) {
			case 'last7Days':
				setDate({
					from: addDays(new Date(), -7),
					to: new Date()
				});
				break;
			case 'last30Days':
				setDate({
					from: addDays(new Date(), -30),
					to: new Date()
				});
				break;
			case 'monthToDate':
				setDate({
					from: new Date(new Date().getFullYear(), new Date().getMonth(), 1),
					to: new Date()
				});
				break;
			case 'yearToDate':
				setDate({
					from: new Date(new Date().getFullYear(), 0, 1),
					to: new Date()
				});
				break;
			case 'thisYear':
				setDate({
					from: new Date(new Date().getFullYear(), 0, 1),
					to: new Date(new Date().getFullYear(), 11, 31)
				});
				break;
			case 'lastYear':
				setDate({
					from: new Date(new Date().getFullYear() - 1, 0, 1),
					to: new Date(new Date().getFullYear() - 1, 11, 31)
				});
				break;
			default:
				break;
		}
	};

	return (
		<div className={cn('inline-flex', className)}>
			<Popover>
				<PopoverTrigger asChild>
					<Button id='date' variant={'outline'} className={cn('w-[260px] justify-start text-left font-normal', !date && 'text-muted-foreground')}>
						<CalendarIcon className='mr-2 h-4 w-4' />
						{date?.from ? (
							date.to ? (
								<>
									{format(date.from, 'LLL dd, y')} - {format(date.to, 'LLL dd, y')}
								</>
							) : (
								format(date.from, 'LLL dd, y')
							)
						) : (
							<span>Pick a date</span>
						)}
					</Button>
				</PopoverTrigger>
				<PopoverContent className='w-auto p-0' align='end'>
					{showInternalPresets ? (
						<>
							<div className='flex'>
								<div className='justify-evenly p-2'>
									<div className='text-muted-foreground' onClick={() => handlePresetSelect('last7Days')}>
										<span className='font-bold underline'>Presets</span>
									</div>
									<div role='button' className='text-sm text-muted-foreground hover:text-primary' onClick={() => handlePresetSelect('last7Days')}>
										Last 7 Days
									</div>
									<div role='button' className='text-sm text-muted-foreground hover:text-primary' onClick={() => handlePresetSelect('last30Days')}>
										Last 30 Days
									</div>
									<div role='button' className='text-sm text-muted-foreground hover:text-primary' onClick={() => handlePresetSelect('monthToDate')}>
										Month to Date
									</div>
									<div role='button' className='text-sm text-muted-foreground hover:text-primary' onClick={() => handlePresetSelect('yearToDate')}>
										Year to Date
									</div>
									<div role='button' className='text-sm text-muted-foreground hover:text-primary' onClick={() => handlePresetSelect('thisYear')}>
										This Year
									</div>
									<div role='button' className='text-sm text-muted-foreground hover:text-primary' onClick={() => handlePresetSelect('lastYear')}>
										Last Year
									</div>
								</div>
								<Calendar initialFocus mode='range' defaultMonth={date?.from} selected={date} onSelect={setDate} numberOfMonths={numberOfMonths} />
							</div>
						</>
					) : (
						<Calendar initialFocus mode='range' defaultMonth={date?.from} selected={date} onSelect={setDate} numberOfMonths={numberOfMonths} />
					)}
				</PopoverContent>
			</Popover>
			{showExternalPresets && (
				<Select
					onValueChange={(value) => {
						setDate(undefined); // Reset range when a preset is selected
						handlePresetSelect(value);
					}}>
					<SelectTrigger className='w-[140px] rounded-l-none'>
						<SelectValue placeholder='Select Range' />
					</SelectTrigger>
					<SelectContent position='popper'>
						<SelectItem value='last7Days'>Last 7 Days</SelectItem>
						<SelectItem value='last30Days'>Last 30 Days</SelectItem>
						<SelectItem value='monthToDate'>Month to Date</SelectItem>
						<SelectItem value='yearToDate'>Year to Date</SelectItem>
						<SelectItem value='thisYear'>This Year</SelectItem>
						<SelectItem value='lastYear'>Last Year</SelectItem>
					</SelectContent>
				</Select>
			)}
		</div>
	);
}
```

Shreypatel13ll commented:
Yes. Making a separate component makes more sense. 

Working on it...

Shreypatel13ll commented:
@Koning3ton 
### I have committed the following changes as suggested
1. **Removed Date Range Picker from existing Date Picker**
2. **Made another UI component named Date Range Picker**
3. **Added to docs**
    - Added CLI installation
    - Added Manual Installation
    - Added Usage

### Images :
![image](https://github.com/shadcn-ui/ui/assets/68627196/7bcef2b6-0790-4945-8b6b-61f722a94997)
![image](https://github.com/shadcn-ui/ui/assets/68627196/d6da06e1-9c83-483d-9d70-7c79185d9dd4)
![image](https://github.com/shadcn-ui/ui/assets/68627196/458519da-ca88-434b-9e25-937a308fe3f1)

**Please suggest further changes and How to write/pass the tests.**

Shreypatel13ll commented:
@Koning3ton any updates?

Shreypatel13ll commented:
@shadcn @bjacobso Any Updates on this?

chrisidakwo commented on apps/www/registry/default/ui/date-range-picker.tsx:
```diff
@@ -0,0 +1,210 @@
+    to: addDays(new Date(2023, 0, 20), 20),
+  })
+
+  const handlePresetSelect = (value: string) => {
```
**Opinion**

Why not leave the actual preset definitions up to the Dev, but rather define a preset interface/type that is then added as an optional array prop to DateRangePickerProps?

Say something like:
```ts
interface DateRangePreset {
  key: string;
  label: string;
  range: {
    from: Date;
    to: Date;
  } 
}

interface DateRangePickerProps extends React.HTMLAttributes<HTMLDivElement> {
  showExternalPresets?: boolean
  showInternalPresets?: boolean
  numberOfMonths?: 2 | 3 | 4 | 5 | 6
  presets?: DateRangePreset[];
}
```

Obviously, you'd loop through the available presets to render appropriately in the `DateRangePicker` component. I'd say that's a better approach than hardcoding preset options (even though that's most likely what a lot of people would go for).

Shreypatel13ll replied:
It's nice to have presets. Added in the latest commit

chrisidakwo commented on apps/www/registry/default/ui/date-range-picker.tsx:
```diff
@@ -0,0 +1,210 @@
+      <Popover>
+        <PopoverTrigger asChild>
+          <Button
+            id="date"
```
Any reason why you gave this an id?

Shreypatel13ll replied:
```suggestion
       

```
No need of it.

Shreypatel13ll replied:
removed in the commit

chrisidakwo commented on apps/www/registry/default/ui/date-range-picker.tsx:
```diff
@@ -0,0 +1,210 @@
+            {date?.from ? (
+              date.to ? (
+                <>
+                  {format(date.from, "LLL dd, y")} -{" "}
```
Why use space character on the left side of the dash character and then use spaced quotes on the right side?

Shreypatel13ll replied:
```suggestion
                 <>
                     {format(date.from, "LLL dd, y")} - {format(date.to, "LLL dd, y")}                
                 </>
```
Somehow my code formatter adds it unintentionally.

chrisidakwo commented:
Nice stuff! I left 3 comments. Maybe it helps.  👍

Shreypatel13ll commented:
### I have committed the following changes as suggested

- Removed `id` from button
- Removed unnecessary `{" "}`
- Added  `presets`

chrisidakwo commented:
@Shreypatel13ll Whether this is finally merged or not, it's good stuff. At least for those looking for a date range picker with presets, this is def a good starting point!

🎉

Shreypatel13ll commented:
@Koning3ton  @chrisidakwo Your guidance and support throughout this process have been invaluable. I truly appreciate your 
effort in reviewing and suggesting changes. Thank you for your input, it has been very helpful in refining this date range picker, and I hope we can see it through to being merged.

Shreypatel13ll commented:
@chrisidakwo any updates?

Shreypatel13ll commented:
@Koning3ton @chrisidakwo @elliot @bjacobso  any progress on this commit?

chrisidakwo commented:
> @Koning3ton @chrisidakwo @elliot @bjacobso any progress on this commit?

 I cannot provide the response you seek as I'm not part of the shadcn team

raunaqss commented:
Adding this comment to make note that the Date Range Picker example produces the following error:

export 'DateRange' (imported as 'DateRange') was not found in 'react-day-picker' (possible exports: Button, CalendarDay, CalendarMonth, CalendarWeek, Caption, CaptionLabel, Chevron, Day, DayButton, DayFlag, DayPicker, Dropdown, DropdownNav, Footer, Month, MonthCaption, MonthGrid, Months, Nav, Option, Root, Row, Select, SelectionState, UI, Week, WeekNumber, WeekNumberHeader, Weekday, Weekdays, Weeks, addToRange, dateLib, dateMatchModifiers, dayPickerContext, defaultLocale, formatCaption, formatDay, formatMonthCaption, formatMonthDropdown, formatWeekNumber, formatWeekNumberHeader, formatWeekdayName, formatYearCaption, formatYearDropdown, getDefaultClassNames, isDateAfterType, isDateBeforeType, isDateInRange, isDateInterval, isDateRange, isDatesArray, isDayOfWeekType, labelCaption, labelDay, labelDayButton, labelGrid, labelGridcell, labelMonthDropdown, labelNav, labelNext, labelPrevious, labelWeekNumber, labelWeekNumberHeader, labelWeekday, labelYearDropdown, rangeIncludesDate, useDayPicker, useNavigation)

Shreypatel13ll commented:
Try using the code below. I just tried it, and it works for me.
#### Here's a short guide:

## Install the dependences.
1.  `npx shadcn@latest add calendar` [Calendar](https://ui.shadcn.com/docs/components/calendar)
2. `npx shadcn@latest add select` [Select](https://ui.shadcn.com/docs/components/select)
3. `npx shadcn@latest add popover` [Popover](https://ui.shadcn.com/docs/components/popover)
4. `npm i date-fns` [date-fns](https://date-fns.org/)
5. `npm i react-day-picker` [react-day-picker](https://daypicker.dev/)

## Component code:
```
"use client"

import * as React from "react"
import { CalendarIcon } from "@radix-ui/react-icons"
import { format, startOfMonth, startOfYear, subDays } from "date-fns"
import { DateRange } from "react-day-picker"

import { cn } from "@/lib/utils"
import { Button } from "./button"
import { Calendar } from "./calendar"
import {
  Popover,
  PopoverContent,
  PopoverTrigger,
} from "./popover"
import {
  Select,
  SelectContent,
  SelectItem,
  SelectTrigger,
  SelectValue,
} from "./select"

interface DateRangePreset {
  label: string
  range: {
    from: Date
    to?: Date
  }
}

interface DateRangePickerProps extends React.HTMLAttributes<HTMLDivElement> {
  showExternalPresets?: boolean
  showInternalPresets?: boolean
  numberOfMonths?: 2 | 3 | 4 | 5 | 6
  presets?: DateRangePreset[]
}

export function DateRangePicker({
  showExternalPresets,
  showInternalPresets,
  numberOfMonths = 2,
  presets,
  className,
}: DateRangePickerProps) {
  const defaultPresets: DateRangePreset[] = [
    {
      label: "Today",
      range: {
        from: new Date(),
        to: new Date(),
      },
    },
    {
      label: "Yesterday",
      range: {
        from: subDays(new Date(), 1),
        to: subDays(new Date(), 1),
      },
    },
    {
      label: "Last 7 Days",
      range: {
        from: subDays(new Date(), 6),
        to: new Date(),
      },
    },
    {
      label: "Last 30 Days",
      range: {
        from: subDays(new Date(), 29),
        to: new Date(),
      },
    },
    {
      label: "Month to Date",
      range: {
        from: startOfMonth(new Date()),
        to: new Date(),
      },
    },
    {
      label: "Year to Date",
      range: {
        from: startOfYear(new Date()),
        to: new Date(),
      },
    },
    {
      label: "This Year",
      range: {
        from: new Date(new Date().getFullYear(), 0, 1),
        to: new Date(),
      },
    },
  ]

  const allPresets = presets ?? defaultPresets

  const [date, setDate] = React.useState<DateRange | undefined>()

  const handlePresetSelect = (preset: DateRangePreset) => {
    setDate({
      from: preset.range.from,
      to: preset.range.to,
    })
  }

  return (
    <div className={cn("inline-flex", className)}>
      <Popover>
        <PopoverTrigger asChild>
          <Button
            variant={"outline"}
            className={cn(
              "w-[260px] justify-start text-left font-normal",
              !date && "text-muted-foreground",
              showExternalPresets && "rounded-r-none"
            )}
          >
            <CalendarIcon className="mr-2 h-4 w-4" />
            {date?.from ? (
              date.to ? (
                <>
                  {format(date.from, "LLL dd, y")} - {format(date.to, "LLL dd, y")}
                </>
              ) : (
                format(date.from, "LLL dd, y")
              )
            ) : (
              <span>Pick a date</span>
            )}
          </Button>
        </PopoverTrigger>
        <PopoverContent className="w-auto p-0" align="end">
          {showInternalPresets ? (
            <>
              <div className="flex">
                <div className="justify-evenly p-2">
                  <div className="text-muted-foreground">
                    <span className="font-bold underline">Presets</span>
                  </div>
                  {allPresets.map((preset) => (
                    <div
                      key={preset.label}
                      role="button"
                      className="text-sm text-muted-foreground hover:text-primary"
                      onClick={() => handlePresetSelect(preset)}
                    >
                      {preset.label}
                    </div>
                  ))}
                </div>
                <Calendar
                  initialFocus
                  mode="range"
                  defaultMonth={date?.from}
                  selected={date}
                  onSelect={setDate}
                  numberOfMonths={numberOfMonths}
                />
              </div>
            </>
          ) : (
            <Calendar
              initialFocus
              mode="range"
              defaultMonth={date?.from}
              selected={date}
              onSelect={setDate}
              numberOfMonths={numberOfMonths}
            />
          )}
        </PopoverContent>
      </Popover>
      {showExternalPresets && (
        <Select
          onValueChange={(value) => {
            setDate(undefined)
            const selectedPreset = allPresets.find(
              (preset) => preset.label === value
            )
            if (selectedPreset) {
              handlePresetSelect(selectedPreset)
            }
          }}
        >
          <SelectTrigger className="w-[140px] rounded-l-none border-l-0">
            <SelectValue placeholder="Select Range" />
          </SelectTrigger>
          <SelectContent position="popper">
            {allPresets.map((preset) => (
              <SelectItem key={preset.label} value={preset.label}>
                {preset.label}
              </SelectItem>
            ))}
          </SelectContent>
        </Select>
      )}
    </div>
  )
}
```

## Usage
```tsx
import { DateRangePicker } from "<your path>"
```

```tsx
<DateRangePicker
  showExternalPresets
  showInternalPresets
  numberOfMonths={2}     // 2 is default
  presets={[            //remove presets if not needed. If only `range.from` is set only one date is selected instead of range. 
    {
      label: "Yesterday",
      range: {
        from: new Date(new Date().setDate(new Date().getDate() - 1))
      },
    },
    {
      label: "before 7 days",
      range: {
        from: new Date(new Date().setDate(new Date().getDate() - 7))
      },
    },
  ]}
/>
```

shadcn commented:
This PR has been automatically marked as stale due to one year of inactivity. It will be closed in 7 days unless there’s further input. If you believe this PR is still relevant, please leave a comment or provide updated details. Thank you. (This is an automated message)
